kafka-client/src/main/java/dev/responsive/kafka/api/async/internals/AsyncProcessor.java

```diff
@@ -0,0 +1,456 @@
+    this.connectedStoreBuilders = connectedStoreBuilders;
+
+    if (userProcessor == null && userFixedKeyProcessor == null) {
+      throw new IllegalStateException("Both the Processor and FixedKeyProcessor were null");
+    } else if (userProcessor != null && userFixedKeyProcessor != null) {
+      throw new IllegalStateException("Both the Processor and FixedKeyProcessor were non-null");
+    }
+  }
+
+  @Override
+  public void init(final ProcessorContext<KOut, VOut> context) {
+
+    initFields((InternalProcessorContext<KOut, VOut>) context);
+
+    userProcessor.init(userContext);
+
+    completeInitialization();
+  }
+
+  // Note: we have to cast and suppress warnings in this version of #init but
+  // not the other due to the KOut parameter being squashed into KIn in the
+  // fixed-key version of the processor. However, we know this cast is safe,
+  // since by definition KIn and KOut are the same type
+  @SuppressWarnings("unchecked")
+  @Override
+  public void init(final FixedKeyProcessorContext<KIn, VOut> context) {
+
+    initFields((InternalProcessorContext<KOut, VOut>) context);
+
+    userFixedKeyProcessor.init((FixedKeyProcessorContext<KIn, VOut>) userContext);
+
+    completeInitialization();
+  }
+
+  /**
+   * Performs the first half of initialization by setting all the class fields
+   * that have to wait for the context to be passed in to #init to be initialized.
+   * Initialization itself is broken up into two stages, field initialization and
+   *
+   */
+  private void initFields(
+      final InternalProcessorContext<KOut, VOut> context
+  ) {
+    this.streamThreadName = Thread.currentThread().getName();
+    this.taskId = context.taskId();
+
+    this.originalContext = context;
+    this.streamThreadContext = new StreamThreadProcessorContext<>(originalContext);
+    this.userContext = new AsyncContextRouter<>(logPrefix, taskId.partition(), streamThreadContext);
+
+    this.asyncProcessorName = originalContext.currentNode().name();
+    this.logPrefix = String.format(
+        "stream-thread [%s] [%s-%d] ",
+        streamThreadName, asyncProcessorName, taskId.partition()
+    );
+    this.log = new LogContext(logPrefix).logger(AsyncProcessor.class);
+
+    this.finalizableRecords = new FinalizingQueue(logPrefix);
+
+    this.threadPool = getAsyncThreadPool(context, streamThreadName);
+    this.threadPool.addProcessor(
+        taskId.partition(),
+        originalContext,
+        finalizableRecords
+    );
+  }
+
+  private void completeInitialization() {
+    // Set the user's context to processing mode so it knows to expect any calls after
+    // this point to come from the AsyncThread that is processing the event
+    userContext.startProcessingMode();
+
+    final Map<String, AsyncKeyValueStore<?, ?>> accessedStores =
+        streamThreadContext.getAllAsyncStores();
+
+    verifyConnectedStateStores(accessedStores, connectedStoreBuilders, log);
+
+    registerFlushListenerForStoreBuilders(
```

Comment:
doesn't this need to happen before the stores are created and initialized? which would have already happened by this point?

Reply:
Whatever it is in the latest commit is definitely correct -- I definitely had it backwards in my mind when I first wrote the feature, so it's possible some comments are misleading/out of date, but I'm sure it's correct since the test was failing before I fixed the ordering to what it is now